core/src/ldml/ldml_transforms.cpp

```diff
@@ -7,15 +7,262 @@
 
 #include "ldml_transforms.hpp"
 #include "debuglog.h"
+#include <algorithm>
+#include <string>
 
 #ifndef assert
-#define assert(x) // TODO-LDML
+#define assert(x)  // TODO-LDML
 #endif
 
 namespace km {
 namespace kbp {
 namespace ldml {
 
+element::element(const USet &new_u, KMX_DWORD new_flags)
+    : chr(), uset(new_u), flags((new_flags & ~LDML_ELEM_FLAGS_TYPE) | LDML_ELEM_FLAGS_TYPE_USET) {
+}
+
+element::element(km_kbp_usv ch, KMX_DWORD new_flags)
+    : chr(ch), uset(), flags((new_flags & ~LDML_ELEM_FLAGS_TYPE) | LDML_ELEM_FLAGS_TYPE_CHAR) {
+}
+
+bool
+element::is_uset() const {
+  return (flags & LDML_ELEM_FLAGS_TYPE) == LDML_ELEM_FLAGS_TYPE_USET;
+}
+
+signed char
+element::get_order() const {
+  unsigned char uorder = ((flags & LDML_ELEM_FLAGS_ORDER_MASK) >> LDML_ELEM_FLAGS_ORDER_BITSHIFT);
+  return (signed char)uorder;
+}
+
+signed char
+element::get_tertiary() const {
+  unsigned char uorder = ((flags & LDML_ELEM_FLAGS_TERTIARY_MASK) >> LDML_ELEM_FLAGS_TERTIARY_BITSHIFT);
+  return (signed char)uorder;
+}
+
+bool
+element::is_prebase() const {
+  return flags & LDML_ELEM_FLAGS_PREBASE;
+}
+
+bool
+element::is_tertiary_base() const {
+  return flags & LDML_ELEM_FLAGS_TERTIARY_BASE;
+}
+
+KMX_DWORD
+element::get_flags() const {
+  return flags;
+}
+
+bool
+element::matches(km_kbp_usv ch) const {
+  if (is_uset()) {
+    return uset.contains(ch);
+  } else {
+    return chr == ch;
+  }
+}
+
+int
+reorder_sort_key::compare(const reorder_sort_key &other) const {
+  if (primary < other.primary) {
+    return -1;
+  } else if (primary > other.primary) {
+    return 1;
+  } else if (secondary < other.secondary) {
+    return -1;
+  } else if (secondary > other.secondary) {
+    return 1;
```

Comment:
That's quite a lot of if-conditions there... and that has me concerned, with how low-level text processing tends to be.  We'd like to keep things performant as possible, right?

For starters, consider this approach:

```suggestion
let comparePrimary = primary - other.primary;
let compareSecondary = secondary - other.secondary;
// ...
  if(comparePrimary) {
    return comparePrimary;
  else if(compareSecondary) {
    return compareSecondary;
// 
```

Generalize to all remaining ladder rungs, of course.

There are an awful lot of conditionals here, and branch prediction misses can impact low-level performance.  Just this one change should cut the conditionals in half.

----------

But, if we want something even more optimized in that regard...

What if we instead composited the `compare___` values that would result into a single value for a single conditional check at the end - or never even diff'd them at all?  After some double-checking, I believe I see an alternate way forward along these lines, though it'd probably be best to more thoroughly prove correctness here.  Anyway, the basic idea:

> Attribute: order
>
>    This attribute gives the primary order for the elements in the matched string in the @from attribute. The value is a simple integer between -128 and +127 inclusive, or a space separated list of such integers.

(also `tertiary`)

--------------------------------
| bits 31-24   |  23-16 | 15-8  | bits 7-0   |
|--------|-----------|----------|-----------|
| primary | secondary | tertiary | quaternary |

If we started with the values being recognized as 8-bit signed ints, then converted 'em each to 32-bit signed ints and bit-shifted them left as appropriate... I believe that would work.  It would likely be a close shave in some scenarios, such as `{primary: 1, secondary: -128, ...}` vs `{primary: 0, secondary: 127, ...}`.

To simplify, an example with just the first two ordering elements:

`{primary: 1, secondary: -128, ...}`:
```
  00000001 00000000 // primary
+ 11111111 10000000 // secondary (two's complement -128)
-------------------
  00000000 10000000 // composited
```

`{primary: 0, secondary: 127, ...}`:
```
  00000000 00000000 // primary
+ 00000000 01111111 // secondary
-------------------
  00000000 01111111 // composited
```

Assuming `secondary` and `quarternary` are always limited to the same range, we should be good here.  And I don't see us needing to reorder past a 10 character range, let alone 256 characters, so I'm pretty sure the assumption holds.

So... I think this approach should work without an extra bit buffer.  Even if it feels like a pretty close razor-shave for scenarios like the above.  Still... that allows us to composite, then compare the four main reorder values by a single if-condition, which leaves less surface area for branch-prediction failures.  

We would still need a second condition for the tiebreak-check... but since...

> The quaternary weight is the index of the character in the string.

... we should never actually get that far.

Reply:
There's a fairly good shot the 'compiled' LDML reorder values could even pre-compile a bit mask and just 'bitwise-OR' the secondary and quaternary entries into it at compare-time:

```
primary_ 00000000 tertiary 00000000
```

Then...

```
// reorder - ordinal
// initialized with the pre-compiled bitmask from above
reordinal = reordinal | (secondary << 16) | quaternary;
```

Reply:
Definitely some good optimization on the sortkeys. I suspect the element matching will take more time at present, and still need to do segmentation into runs. I'll look at incorporating these, thanks.